File: PracticeProblems/CPSolns/LeetCodeDSI/searchBST.cpp
#include<iostream>
#include "./treeStructure.cpp"
#include "./dfTreeTraversal.cpp"

using std::cout;
using std::endl;

/* 
 * PROBLEM STATEMENT:
 * Given a root of a binary search tree and a target value.
 * Find the node in the BST equal to the target and return subtree rooted with that node
 * otherwise return false.
 *
 * Sample:
 * Input: root = [4, 2, 7, 1, 3], target = 2                4
 * Output: [2, 1, 3]                                       /  \
 *                                                         2   7
 *                                                        /  \
 *                                                        1   3
 *
 *
 * Approach:
 * In a bst, values less than root are stored in left side and value greater than root 
 * are stored in right side.
 * If target is greater than root traverse the left tree, once found return the node
 * Otherwise traverse the left subtree.
 */

class Solution{
    public:
        TreeNode* searchBstRec(TreeNode* root, int target){
            if(!root) return {};
            if(root->data == target)
                return root;
            else if(root->data > target)
                return searchBstRec(root->left, target);
            else
                return searchBstRec(root->right, target);
        }

        TreeNode* searchBstIter(TreeNode* root, int target){
            if(!root) return {};
            if(root->data == target)
                return root;
            else if(root->data > target)
                root = root->left;
            else
                root = root->right;
            return root;
        }
};

int main(){
    Solution soln;
    TraverseTree traverse;
    TreeNode* root = newNode(4);
    root->left = newNode(2);
    root->right = newNode(7);
    root->left->left = newNode(1);
    root->left->right = newNode(3);
    TreeNode* resRoot = soln.searchBstRec(root, 2);
    vector<int> resRec = traverse.preOrder(resRoot);
    traverse.printArr(resRec);
    return 0;   
}

File: Datastructures/PracticeProblems/LeetCodeDSI-JavaSolns/SearchBST.java
// Give root of binary search tree and target, return subtree rooted with the target node
// otherwise return null
// Leet code 700

import java.util.List;

class SearchBST{
    static TreeNode root;
    static TreeNode root1;

    SearchBST(int data){
        root = new TreeNode(data);
    }

    SearchBST(){
        root = null;
    }

    static TreeNode searchBst(TreeNode root, int target){
        if(root == null) return null;
        while(root != null){
            if(root.data > target)
                root = root.left;
            else if(root.data < target)
                root = root.right;
            else
                return root;
        }
        return null;
    }

    void printList(List<Integer> arr){
        for(Integer i : arr)
            System.out.print(i + " ");
        System.out.println();
    }

    public static void main(String[] args){
        LevelOrderTraversal lOrder = new LevelOrderTraversal();
        SearchBST tree = new SearchBST();
        tree.root = new TreeNode(4);
        root.left = new TreeNode(2);
        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);
        root.right = new TreeNode(7);

        List<List<Integer>> treeList = lOrder.levelOrder(root);
        System.out.println("Tree");
        lOrder.printList(treeList);
        System.out.println("target: 2");

        tree.root1 = searchBst(root, 2);
        List<List<Integer>> subTreeList = lOrder.levelOrder(root1);
        System.out.println("Found sub tree");
        lOrder.printList(subTreeList);
    }
}


